### PR TITLE
Editor: add local variables to watch panel automatically

### PR DESCRIPTION
**Pull request by ericoporto**
This change assumes local variables are usually meaningful in the context of the function being debugged.  
The user can still add and remove variables to watch, for meaningful global variables.

This is a draft that perhaps should have been an issue, but anyway, I wanted to show to see what is useful or not from this. The limitation here is the `GetListOfLocalVariablesForCurrentPosition` function, which relies on parsers in both AutoComplete and ScintillaWrapper which aren't perfect - perhaps they could be improved with more testing.

https://github.com/adventuregamestudio/ags/assets/2244442/090e668a-e144-4c76-8884-76fd73d0fc32

Made a fix for the blinking in the video above please try and see what you think of this change.

https://github.com/adventuregamestudio/ags/assets/2244442/787b2977-b0d1-4abb-b98b-2ab1863b8f26

### CODE REVIEW

**Comment by ericoporto** [comment]
Added a very minimal left click context menu to add things to watch pane

image

**Comment by ivan-mogilko** [comment]
Sorry for a such a long delay, I'd like to review and deal with this PR draft now if possible.
@ericoporto could you please rebase this, it seems to have few conflicts with the recent ags4 branch?

**Comment by ericoporto** [comment]
Right, I will do it tonight. The reason I never took it off of draft status is it had an issue that it would occasionally lock the AGS game when I ran. After I rebase, if I find the issue again I will mention it.

**Comment by ericoporto** [comment]
Hey, I did the rebase but I haven't been able to test it yet. :/

**Comment by ivan-mogilko** [comment]
There's something funny, having `new Type` anywhere in the code would add a local variable "Type" entry (of course it won't be resolved).

For example:
```
int arr[] = new int[1];
```

Not sure which part of this feature is causing this.

I'll continue testing.

**Comment by ivan-mogilko** [comment]
So, I noticed that this method `GetListOfLocalVariablesForCurrentPosition` requires ScintillaWrapper, and as a consequence you have to get ScriptEditor control. While we may assume that a active script's control should be available when you hit a breakpoint, I think this should be avoided.

Looking at GetListOfLocalVariablesForCurrentPosition, it seems like this method does not have to be in ScintillaWrapper at all, because it does not access any of scintilla's functionality, but only AutoComplete functionality. It may be possible to move this method to some Utilities class, where it would accept only IScript, and work with its Text and AutoCompleteData directly.

**Comment by ivan-mogilko** [comment]
Overall, I think that this is a pretty good PR. This extra feature is achieved with a small amount of code, and is rather straightforward. The only questionable thing there is how local variables are scanned every time, but they are limited by a single function's scope, and it's rather the problem of autocomplete system. IMO it's a shame that AutoComplete does not cache local variables as it does globals. But this may be improved separately.

Something that bothers me slightly in the code is that much stuff is done under `_updateItemLock` when syncing local vars, usually it's a good thing to minimize time spent under a lock by doing everything that does not have to be protected before or after. Maybe this code could be reorganized. OTOH I found a place in my earlier code that also could be improved for a similar purpose (one where it calls `UpdateSingleWatch` in a loop).

---

I suggest adding a checkable option in the panel's context menu that turns automatic local variables on and off, in case a user does not want to see them.

---

"Add to Watch" menu command is a very nice addition, but it has couple of bugs:
1. It is enabled for any symbols not just variables,
2. When a variable gets added to the list this way, there's an extra empty item inserted before for some reason.

**Comment by ericoporto** [comment]
Thanks for the review, did you not get the issue where ags occasionally locks itself? That is the biggest issue for me, I couldn't progress much because I kept hitting the locking issue, it happens when debugging a game, I noticed in my Sandwalker game but I also got it in others, if I kept using the step function it would eventually lock the game - alt+tab to it would find a hang window.

**Comment by ivan-mogilko** [comment]
> did you not get the issue where ags occasionally locks itself? That is the biggest issue for me, I couldn't progress much because I kept hitting the locking issue, it happens when debugging a game, I noticed in my Sandwalker game but I also got it in others, if I kept using the step function it would eventually lock the game - alt+tab to it would find a hang window.

No, but I did not test for too long.
Does it happen in random moments, or at particular places?

To clarify, this locks the running game, not the editor, correct?

**Comment by ericoporto** [comment]
Correct, the Editor keeps working, it's only the game. I will make a video to show, because it's weird, and Windows gets unhappy that the game window is hang so it does that thing where it shows the "it has stopped working" message and gives you the option to kill it.

**Comment by ivan-mogilko** [comment]
I suppose this may be debugged by attaching another instance of MSVS to a running game process.

File: Editor/AGS.Editor/Panes/ScriptEditorBase.cs [hunk 1]
```diff
@@ -97,7 +97,7 @@ protected IScript Script
         /// <summary>
         /// Lets for a child class to assign an actual Scintilla control.
         /// </summary>
-        protected ScintillaWrapper Scintilla
+        public ScintillaWrapper Scintilla
```
[the remark this answers is not in the document]
Comment (ericoporto) [reply Editor/AGS.Editor/Panes/ScriptEditorBase.cs]: this is correct, this is not needed, this is a left over from a previous experiment

**Comment by ericoporto** [comment]
> I suggest adding a checkable option in the panel's context menu that turns automatic local variables on and off, in case a user does not want to see them.

Uhm, this is an interesting idea, I tried one way of making it here

https://github.com/ericoporto/ags/commit/95656002bc910fee0973de5ba5424f8052e1351c

The problem is... There is a bug in Windows 11 that the checkbox scales weirdly and may not be visible in context menus... (I think the bug is reported here https://github.com/dotnet/winforms/issues/9258). Anyway, in my PC the checkbox is sort of hard to understand it's a checkbox.

I think this could be put in the Editor Preferences and be done, and if in the future we have more needs for the watchpanel we may add some toolbar in it like the log panel.

**Comment by ivan-mogilko** [comment]
> The problem is... There is a bug in Windows 11 that the checkbox scales weirdly and may not be visible in context menus... (I think the bug is reported here https://github.com/dotnet/winforms/issues/9258). Anyway, in my PC the checkbox is sort of hard to understand it's a checkbox.

I do not think that this is a good reason to *not* add an option into context menu. That's cosmetic issue vs convenient control element. Besides, we may have a need for other checkable items in menus. EDIT: there's already a checkable option called "Word Wrap" in the Edit menu in ags4.

I see following alternatives here:
1. Just add it, and tolerate that it does not look well on some systems until there's a fix.
2. Find a way to draw checkable items ourselves (do those menus allow to override paint operation?).
3. Add 2 options, one of which enables and another disables this mode, and one of them is either disabled or hidden depending on current choice.

**Comment by ivan-mogilko** [comment]
>  and if in the future we have more needs for the watchpanel we may add some toolbar in it like the log panel.

Ah yes, toolbar is another good alternative.

**Comment by ivan-mogilko** [comment]
I found that it's very easy to reproduce the hanging problem: just place a breakpoint inside a repeating script, and hold F5. It happens in a split second.

My impression is that this is because _communicator.SendMessage is being called from multiple threads and its private members  are not being protected by any lock. But other than that, the engine reading debugger messages is written with assumption that there's only one message at a time, but probably that's not guaranteed. Normally it should parse received string piece by piece, until it is all read out.

Look like this is a general issue. It does not occur often maybe because it requires a big chain of messages to be sent simultaneously.

**Comment by ivan-mogilko** [comment]
@ericoporto I created a commit in my repo here: https://github.com/ivan-mogilko/ags-refactoring/commit/145d5a921e23e57f59911a54e724792b005ed240

This seems to fix the engine getting locked in the waiting state.

**Comment by ericoporto** [comment]
I confirm that fix makes the error go away!!!!

**Comment by ericoporto** [comment]
> - local variables are added automatically only if this option is enabled;
> - cannot be removed by Remove or Clear menu commands;
> - removed completely if user unchecks the option.

I believe I did these.

> - automatically added variables are always on top of the list;

I haven't yet done this, still haven't figured it out, I will try again tomorrow.

**Comment by ivan-mogilko** [comment]
> automatically added variables are always on top of the list;
I haven't yet done this, still haven't figured it out

Is not that how it currently works? Last time I checked the code, i noticed that you are inserting local variables at index 0 in the list.

**Comment by ericoporto** [comment]
I thought it was a task list but I couldn't figure it what I did wrong.

**Comment by ericoporto** [comment]
I think I fixed the problems in the new features of auto-local vars and add to watch!

 > It is enabled for any symbols not just variables,

I don't plan to fix this in this PR, couldn't find a reliable way to filter for only variables.

**Comment by ivan-mogilko** [comment]
Technically it is all working. I found several minor issues.

1. The Watch Panels' context menu now looks different from other menus, because the "check" icon has non-standard size.
Here's the comparison between script's context menu and watch panel's:
watchpanel-contextcomparison


2. I'd suggest moving "Add to watch panel" command lower, because "Go to definition" and "Find all usages" seem logically connected commands.
3. I noticed that if there's nothing useful under the cursor, the "Add to Watch Panel" command is disabled, but the text is truncated to just "Add", which looks weird. Is there's a way to make it "Add to Watch Panel" (without any name in the middle), for consistency with "Go to Definition"?
4. I just got an idea that it would be cool if we could drag & drop a word from a script to watch panel. Maybe later.
5. If I enable "Autowatch local variables" while at a breakpoint, the variables are not added to the list until I make a single step.

**Comment by ericoporto** [comment]
I believe I did 1, 2, and 3. I haven't figured 5, and leaving 4 for the future.

**Comment by ericoporto** [comment]
OK, did 1, 2, 3 and 5.

> I just got an idea that it would be cool if we could drag & drop a word from a script to watch panel. Maybe later.

This sounds really cool, but I think I would need to mess with the script editor base. The reason is if you actually do a drop - I hacked a version of this - it deletes the word from the script once you handle it where it drops.

**Comment by ericoporto** [comment]
OK, I managed to implement 1, 2, 3, 4 and 5!

This ended up being 11 different commits, should I squash everything or better leave them as they are? The way it is I think at least documents a little why each piece of the code was added, at least per features/issues.

**Comment by ericoporto** [comment]
> @ericoporto I created a commit in my repo here: [ivan-mogilko@145d5a9](https://github.com/ivan-mogilko/ags-refactoring/commit/145d5a921e23e57f59911a54e724792b005ed240)
> 
> This seems to fix the engine getting locked in the waiting state.
> 
> Please try if that works for you?

Just thinking about this now, in theory could this locking issue happen in ags3 too? Do we need to backport this? I didn't get the issue in ags3 before but perhaps I haven't used the debug a lot there since there isn't a watch panel in it.

**Comment by ivan-mogilko** [comment]
> Just thinking about this now, in theory could this locking issue happen in ags3 too? Do we need to backport this? I didn't get the issue in ags3 before but perhaps I haven't used the debug a lot there since there isn't a watch panel in it.

I don't think that you can get locking there in practice, because all commands are issued from the same thread.
But it may be backported of course.

**Comment by ericoporto** [comment]
Ah, ok, if it's not a problem there is no need. Thanks!

**Pull request merged**